File: backend/token_estimate.py
"""Model-specific token heuristics (parity with frontend lib/tokenEstimate.js)."""

TOKEN_MODELS = ("GPT-4", "Claude", "LLaMA")

_MULTIPLIERS = {
    "GPT-4": 1.3,
    "Claude": 1.2,
    "LLaMA": 1.4,
}


def _word_count(text: str) -> int:
    t = text.strip()
    if not t:
        return 0
    return len(t.split())


def estimate_tokens_by_model(text: str, model: str) -> float:
    mult = _MULTIPLIERS.get(model, _MULTIPLIERS["GPT-4"])
    return round(_word_count(text) * mult, 1)

File: frontend/lib/tokenEstimate.js
/**
 * Model-specific token heuristics (frontend-only).
 * GPT: word × 1.3 (tiktoken-quality counting would need WASM/server; this matches prior app default).
 * Claude: word × 1.2
 * LLaMA: word × 1.4
 */

export const TOKEN_MODELS = ["GPT-4", "Claude", "LLaMA"];

const MULTIPLIERS = {
  "GPT-4": 1.3,
  Claude: 1.2,
  LLaMA: 1.4,
};

function wordCount(text) {
  return text.trim() ? text.trim().split(/\s+/).length : 0;
}

/**
 * @param {string} text
 * @param {string} model One of TOKEN_MODELS
 */
export function estimateTokensByModel(text, model) {
  const mult = MULTIPLIERS[model] ?? MULTIPLIERS["GPT-4"];
  return Math.round(wordCount(text) * mult * 10) / 10;
}
